### PR TITLE
HDDS-6358. EC: Refactor ECKeyOutputStream#write()

### PR DESCRIPTION
**Pull request by kaijchen**
## What changes were proposed in this pull request?

Refactor the code to make it cleaner, here is a summary of the changes.

1. Instead of treating `head/middle/tail` differently, use a general way to write data into buffers.
2. No more `allocateBlockIfNeeded` parameters being passed around.
3. Stream will be closed if any exception happens, and `IOException` will be thrown.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-6358

## How was this patch tested?

Unit tests.

### CODE REVIEW

**Comment by guihecheng** [comment]
Hi @kaijchen the overall patch LGTM!
It's better to have several lines of summary about the changes made here to be easier for reviewing, I believe that there are certain points to be mentioned that is more than cleanup.
cc @umamaheswararao

**Comment by kaijchen** [comment]
@guihecheng Thanks for reviewing, I have edited the PR message to include a summary of changes.

**Comment by umamaheswararao** [comment]
Thanks for filing this @kaijchen
Initially journey was started with this single [loop | https://github.com/apache/ozone/blob/f9a0782ae4efebabb55127df3954745a3c4a0889/hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/io/ECKeyOutputStream.java] 
However for easy trace tracking at which position write impacted, I have decided to split into 3 sections as you got that in summary comments. 
I think now, we have evolved enough and bit stable compared to starting point where we need lot of debug help info to track down issues. ( When it has 3 sections I found helpful as the trace indicates which position of data sizes it's causing issues. We may need to add more debug info as well to improve debug-ability.)

Overall patch looks good to me. I will pass one more round as this is in key path and provide my feedback if any.

Thanks a lot for working on this.

File: hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/io/ECKeyOutputStream.java [hunk 1]
```diff
@@ -540,26 +515,23 @@ public void close() throws IOException {
   }
 
   private void handleStripeFailure(int lastStripeSize,
-      boolean allocateBlockIfFull, boolean isClose)
-      throws IOException {
+      boolean isClose) throws IOException {
```
Comment (umamaheswararao) [inline hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/io/ECKeyOutputStream.java:518]: After var removal, this lines may need format?

File: hadoop-ozone/client/src/test/java/org/apache/hadoop/ozone/client/TestOzoneECClient.java [hunk 2]
```diff
@@ -573,8 +573,8 @@ public void testStripeWriteRetriesOn2Failures() throws IOException {
         nodesIndexesToMarkFailure);
```
Comment (umamaheswararao) [inline hadoop-ozone/client/src/test/java/org/apache/hadoop/ozone/client/TestOzoneECClient.java:573]: I find all of these changes are unrelated as well. I would suggest to repurpose other JIRA as cleanups task and do this changes. Hope that make sense to you.

**Review by umamaheswararao — COMMENTED** [review]
@kaijchen, Thanks for working on this task. I have dropped few comments. Please check.

File: hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/io/ECKeyOutputStream.java [hunk 3]
```diff
@@ -157,57 +155,47 @@ public void write(byte[] b, int off, int len) throws IOException {
...
-      checkAndWriteParityCells(pos, iters > 0 || remLen > 0);
-    }
-
-    if (lastCellSize > 0) {
-      currentStreamIdx = blockOutputStreamEntryPool.getCurrentStreamEntry()
-          .getCurrentStreamIdx();
-      pos = handleDataWrite(currentStreamIdx, b, off,
-          lastCellSize, false);
-      checkAndWriteParityCells(pos, false);
+  private void handleWrite(byte[] b, int off, long len, boolean retry)
```
[the remark this answers is not in the document]
Comment (kaijchen) [reply hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/io/ECKeyOutputStream.java:162]: I think we can probably override KeyOutputStream#handleWrite().

File: hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/io/ECKeyOutputStream.java [hunk 4]
```diff
@@ -157,57 +155,47 @@ public void write(byte[] b, int off, int len) throws IOException {
     if (len == 0) {
       return;
     }
-    blockOutputStreamEntryPool.allocateBlockIfNeeded();
+    handleWrite(b, off, len, false);
+    writeOffset += len;
+  }
```
[the remark this answers is not in the document]
Comment (kaijchen) [reply hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/io/ECKeyOutputStream.java:161]: Can we make KeyOutputStream#handleWrite() protected and override it?

**Comment by kaijchen** [comment]
Hi @umamaheswararao, thanks for reviewing. I have reverted most of the unrelated changes.
Can we override `KeyOutputStream#handleWrite()` and remove the `write()` here?
Because currently `ECKeyOutputStream#write()` is exactly same as `KeyOutputStream#write()`.

File: hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/io/ECKeyOutputStream.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (kaijchen) [reply hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/io/ECKeyOutputStream.java:162]: If we make KeyOutputStream#handleWrite() protected. We can override it here.

File: hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/io/ECKeyOutputStream.java [hunk 5]
```diff
@@ -157,57 +155,47 @@ public void write(byte[] b, int off, int len) throws IOException {
...
-    if (lastCellSize > 0) {
-      currentStreamIdx = blockOutputStreamEntryPool.getCurrentStreamEntry()
-          .getCurrentStreamIdx();
-      pos = handleDataWrite(currentStreamIdx, b, off,
-          lastCellSize, false);
-      checkAndWriteParityCells(pos, false);
+  private void handleWrite(byte[] b, int off, long len, boolean retry)
+      throws IOException {
+    while (len > 0) {
+      try {
```
[the remark this answers is not in the document]
Comment (kaijchen) [reply hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/io/ECKeyOutputStream.java:165]: Yes. But will there be noticeable performance difference? Because `KeyOutputStream` is doing the same thing.
If we need to add this check, can we add it inside `allocateBlockIfNeeded()`?

**Comment by kaijchen** [comment]
Hi @umamaheswararao, thanks for the explanation and the review.
I have updated the patch to resolve most of the comments. And there is one needs further discussion.
I think we can let block allocation happen here instead of inside `handleDataWrite` or `handleParityWrite`.
What do you think?

File: hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/io/ECKeyOutputStream.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (umamaheswararao) [reply hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/io/ECKeyOutputStream.java:157]: Few if checks might not contribute to significant improvement, but this is key path and saving such thing also helps. But doing that checks needs to modify BlockOutputStreamEntryPool, that should be done in master I feel as that is common. So, I am ok to keep it here now as it is.

**Review by umamaheswararao — APPROVED** [review]
LGTM, pending CI
